plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
+/* tslint:disable:no-any */
+import { State } from '@hookstate/core';
+import { PluginCallbacks, StateValueAtRoot } from '@hookstate/core/dist';
 
-import { Plugin, Path, StateValueAtPath, State } from '@hookstate/core';
+export const ValidationId = Symbol('Validation');
 
-export type ValidationSeverity = 'error' | 'warning';
+type ValidateFn<T> = (value: T) => boolean;
+type Path = readonly (string | number | symbol)[];
 
-export interface ValidationError {
-    readonly message: string;
-    readonly path: Path;
-    readonly severity: ValidationSeverity;
+interface NestedState {
+    state: State<any>;
+    parent?: NestedState;
 }
 
-export interface ValidationExtensions<S> {
-    validate(attachRule: (value: S) => boolean,
-        message: string | ((value: S) => string),
-        severity?: ValidationSeverity): void,
-    validShallow(): boolean,
-    valid(): boolean,
-    invalidShallow(): boolean,
-    invalid(): boolean,
-    firstError(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number
-    ): Partial<ValidationError>,
-    errors(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number,
-        first?: boolean
-    ): ReadonlyArray<ValidationError>,
+interface SingleValidator<T> {
+    required(message?: string): void;
+
+    when(fn: (value: State<T>) => boolean): this;
+
+    validate(validator: ValidateFn<T>, message?: string): void;
 }
 
-const PluginID = Symbol('Validate');
+type NestedValidator<Root, T> = T extends string ? SingleValidator<T> :
+    T extends any[] ? ArrayValidator<Root, T[0]> : ObjectValidator<Root, T>;
 
-const emptyErrors: ValidationError[] = []
+type ObjectNestedValidator<Root, T> = {
+    [Key in keyof T]: NestedValidator<Root, T[Key]>
+};
 
-interface ValidationRule {
-    readonly message: string | ((value: StateValueAtPath) => string)
-    readonly rule: (v: StateValueAtPath) => boolean
-    readonly severity: ValidationSeverity;
+interface ObjectRootValidator<Root, T> extends SingleValidator<T> {
+    when(fn: (value: State<T>, root: Root) => boolean): this;
 }
 
-class ValidationPluginInstance<S> {
-    private storeRules = {};
+type ObjectValidator<Root, T> = ObjectNestedValidator<Root, T> & ObjectRootValidator<Root, T> & {
+    whenType<K extends keyof T, V extends T[K]>(key: K, value: V):
```

Comment:
would be good to see tests with examples. I can not get how to use it...

Reply:
i've updated the PR description with some examples... i'll write some tests soon as i hammer out all my real-world use cases... i keep find new approaches to the API :)